types/neovim/template

```diff
@@ -250,25 +277,121 @@ global record vim
+         source: string -- The source of the diagnostic
+      end
+      record ConfigOpts
+         record UnderlineOpts
+            severity: integer -- vim.diagnostic.severity
+         end
+         record VirtualTextOpts
+            severity: integer -- vim.diagnostic.severity
+            source: string
+            format: function(diagnostic: Diagnostic): string
+         end
+         record SignsOpts
+            severity: integer
+            priority: integer
          end
-         goto_prev: function
-         goto_next: function
-         show_line_diagnostics: function(ShowLineOpts, number, number, number): number, number
+
+         underline: boolean|UnderlineOpts
+         virtual_text: boolean|VirtualTextOpts
+         signs: boolean|SignsOpts
+         float: OpenFloatOpts
+         update_in_insert: boolean
+         severity_sort: boolean
+      end
+      record GetOpts
+         namespace: integer
+         lnum: integer
+         severity: integer -- vim.diagnostic.severity
+      end
+      record GotoOpts
+         namespace: integer
+         cursor_position: {integer, integer}
+         wrap: boolean
+         severity: integer -- vim.diagnostic.severity
+         float: boolean|OpenFloatOpts
+         win_id: integer
       end
+      record OpenFloatOpts
+         namespace: integer
+         scope: string -- "buffer", "line" or "cursor"
+         pos: integer|{integer, integer}
+         severity_sort: boolean
+         severity: integer -- vim.diagnostic.severity
+         header: string|{string, string}
+         source: string
+         format: function(diagnostic: Diagnostic): string
+         prefix: string|{string, string}|function(diagnostic: Diagnostic, i: integer, total: integer): string, string|nil
+      end
+      record SetLocListOpts
+         namespace: integer
+         winnr: integer
+         open: boolean
+         title: string
+         severity: integer -- vim.diagnostic.severity
+      end
+      record SetQfListOpts
+         namespace: integer
+         open: boolean
+         title: string
+         severity: integer -- vim.diagnostic.severity
+      end
+      config: function(opts: ConfigOpts, namespace: integer|nil)
+      disable: function(bufnr: integer|nil, namespace: integer|nil)
+      disable: function(bufnr: integer|nil, namespace: integer|nil)
+      fromqflist: function(list: {any}): {Diagnostic}
+      get: function(bufnr: integer|nil, opts: GetOpts)
+      get_namespace: function(namespace: integer): {string:integer}
+      get_namespace: function(namespace: integer): {{string:integer}}
+      get_next: function(opts: GotoOpts): Diagnostic
+      get_next_pos: function(opts: GotoOpts): {integer, integer}
+      get_prev: function(opts: GotoOpts): Diagnostic
+      get_prev_pos: function(opts: GotoOpts): {integer, integer}
+      goto_next: function(opts: GotoOpts)
+      goto_prev: function(opts: GotoOpts)
+      hide: function(namespace: integer|nil, bufnr: integer|nil)
```

Comment:
Currently everything in Teal is `nil`able by default and until that changes I don't see the annotation as useful

Reply:
So should I remove all of the `|nil` type declarations? I thought I might add them to be more explicit, the docs say:
```
Parameters: ~
    {namespace}  number|nil Diagnostic namespace. When
                 omitted, hide diagnostics from all
                 namespaces.
    {bufnr}      number|nil Buffer number, or 0 for current
                 buffer. When omitted, hide diagnostics in all
                 buffers.
```

Reply:
Yes please, if/when Teal gets non-nilable types (or something like [this](https://github.com/teal-language/tl/pull/366) gets merged) I don't think it will be too hard to just go through each function and re-annotate it.

Personal opinion here, but I personally think that explicitly putting the `|nil` gives a false sense of security that the type checker will catch potential `nil` (which it won't). Maybe on another library this would be useful, but I'd argue if you want documentation you can just use neovim's `:help` to get the info you need. This is just to make the type checker happy.

Reply:
Ok, done